Automattic/newspack-plugin
# feat(data-events): track content gate interactions

### All Submissions:






### Changes proposed in this Pull Request:

1200550061930446-as-1205734635694818

Creates `gate_interaction` data events with the following schema:

| Name           | Type     | Obs                                                                                               |
| -------------- | -------- | ------------------------------------------------------------------------------------------------- |
| `gate_post_id` | `int`    |                                                                                                   |
| `action`       | `string` | Either `seen`, `form_submission_received`, `form_submission_success` or `form_submission_failure` |
| `action_type`  | `string` | Either `paid_membership` or `registration`. Not applicable when `action` is `seen`                |
| `referer`      | `string` |                                                                                                   |
| `order_id`     | `int`    | Only applicable when `action_type` is `paid_membership`                                           |
| `product_id`   | `int`    | Only applicable when `action_type` is `paid_membership`                                           |
| `amount`       | `float`  | Only applicable when `action_type` is `paid_membership`                                           |
| `currency`     | `string` | Only applicable when `action_type` is `paid_membership`                                           |

To allow tracking form interactions inside the gate, a script injects a hidden `memberships_content_gate` input to forms inside the gate. This is passed by the reader registration or the modal checkout logic so it can finally get picked up by the data event listeners.

It also implements the integration with the ga4 connector, which will include additional parameters to the GA event: `logged_in`, `is_reader`, and `email_hash`.

### How to test the changes in this Pull Request:

1. Checkout this branch and https://github.com/Automattic/newspack-blocks/pull/1590
2. Make sure you have GA4 connected with the wp options `ga4_measurement_id` and `ga4_measurement_protocol_secret` configured
3. Set up a membership plan restricted by user registration and configure an inline content gate with a reg. wall (reader registration block)
4. Go through the gate anonymously and make sure your session allows 3rd party cookies
5. Visit GA's realtime dashboard, wait until the events are sent and confirm you get 3 `np_gate_interaction` events: `seen` and `form_submission_received` and `form_submission_success` with `action_type` set to `registration` and the correct `referer`
6. Change the membership plan to grant access through a subscription product
7. Change the content gate from "inline" to the "overlay" style
8. Repeat steps 3 and 4 but confirm that this time the `action_type` is `paid_membership`
9. Confirm the event also includes `order_id`, `product_id`, `amount` and `currency`

### Other information:
